clients/src/main/java/org/apache/kafka/clients/ClientTelemetryReporter.java

```diff
@@ -0,0 +1,937 @@
+
+    @SuppressWarnings("unchecked")
+    @Override
+    public synchronized void configure(Map<String, ?> configs) {
+        rawOriginalConfig = (Map<String, Object>) Objects.requireNonNull(configs);
+    }
+
+    @Override
+    public synchronized void contextChange(MetricsContext metricsContext) {
+        /*
+          If validation succeeds: initialize the provider, start the metric collection task,
+          set metrics labels for services/libraries that expose metrics.
+         */
+        Objects.requireNonNull(rawOriginalConfig, "configure() was not called before contextChange()");
+        collectors.forEach(MetricsCollector::stop);
+
+        if (!telemetryProvider.validate(metricsContext)) {
+            log.warn("Validation failed for {} context {}, skip starting collectors. Metrics collection is disabled",
+                    telemetryProvider.getClass(), metricsContext.contextLabels());
+            return;
+        }
+
+        if (kafkaMetricsCollector == null) {
+            // Initialize the provider only once. contextChange(..) can be called more than once,
+            // but once it's been initialized and all necessary labels are present then we don't
+            // re-initialize.
+            telemetryProvider.configure(rawOriginalConfig);
+        }
+
+        telemetryProvider.contextChange(metricsContext);
+
+        if (kafkaMetricsCollector == null) {
+            initCollectors();
+        }
+    }
+
+    @Override
+    public void init(List<KafkaMetric> metrics) {
+        // metrics collector may not have been initialized (e.g. invalid context labels)
+        // in which case metrics collection is disabled
+        if (kafkaMetricsCollector != null) {
+            kafkaMetricsCollector.init(metrics);
+        }
+    }
+
+    /**
+     * Method is invoked whenever a metric is added/registered
+     */
+    @Override
+    public void metricChange(KafkaMetric metric) {
+        // metrics collector may not have been initialized (e.g. invalid context labels)
+        // in which case metrics collection is disabled
+        if (kafkaMetricsCollector != null) {
+            kafkaMetricsCollector.metricChange(metric);
+        }
+    }
+
+    /**
+     * Method is invoked whenever a metric is removed
+     */
+    @Override
+    public void metricRemoval(KafkaMetric metric) {
+        // metrics collector may not have been initialized (e.g. invalid context labels)
+        // in which case metrics collection is disabled
+        if (kafkaMetricsCollector != null) {
+            kafkaMetricsCollector.metricRemoval(metric);
+        }
+    }
+
+    @Override
+    public void close() {
+        log.debug("Stopping ClientTelemetryReporter");
+        try {
+            clientTelemetrySender.close();
+        } catch (Exception exception) {
+            log.error("Failed to close client telemetry reporter", exception);
+        }
+    }
+
+    public synchronized void updateMetricsLabels(Map<String, String> labels) {
```

Comment:
I cannot find where this is called, and thus it's not clear to me why it's added (similar for `telemetryProvider.updateLabels(...)` which is only called from here as it seems)

Reply:
The usage is in the PR: https://github.com/apache/kafka/pull/14843/files to update group_member_id metric label.